src/DefaultBuilder/src/BootstrapHostBuilder.cs

```diff
@@ -1,36 +1,49 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
 using Microsoft.Extensions.Configuration;
 using Microsoft.Extensions.DependencyInjection;
-using Microsoft.Extensions.FileProviders;
 using Microsoft.Extensions.Hosting;
 
 namespace Microsoft.AspNetCore.Hosting;
 
 // This exists solely to bootstrap the configuration
 internal sealed class BootstrapHostBuilder : IHostBuilder
 {
-    private readonly IServiceCollection _services;
+    private readonly HostApplicationBuilder _builder;
+
     private readonly List<Action<IConfigurationBuilder>> _configureHostActions = new();
     private readonly List<Action<HostBuilderContext, IConfigurationBuilder>> _configureAppActions = new();
     private readonly List<Action<HostBuilderContext, IServiceCollection>> _configureServicesActions = new();
 
-    private readonly List<Action<IHostBuilder>> _remainingOperations = new();
-
-    public BootstrapHostBuilder(IServiceCollection services, IDictionary<object, object> properties)
+    public BootstrapHostBuilder(HostApplicationBuilder builder)
     {
-        _services = services;
+        _builder = builder;
 
-        Properties = properties;
+        foreach (var descriptor in _builder.Services)
```

Comment:
I thought the goal was to remove the hacks...

Reply:
> I hope to propose a better public API than ConfigureWebHostDefaults to consume the default configuration from Microsoft.Extensions.Hosting in Microsoft.AspNetCore, but that will likely have a big surface area that will need API reviewed. I think this can be done as a follow up.

Reply:
https://github.com/dotnet/aspnetcore/issues/40761